docs/guides/single-fetch.md

```diff
@@ -98,13 +98,66 @@ In order to ensure you get the proper types when using Single Fetch, we've inclu
 
 ```json
 {
-  "include": [
-    // ...
-    "node_modules/@remix-run/react/future/single-fetch.d.ts"
-  ]
+  "compilerOptions": {
+    "types": ["@remix-run/react/future/single-fetch.d.ts"]
+  }
 }
 ```
 
+🚨 Make sure the single-fetch types come after any other Remix packages in `types` so that they override those existing types.
+
+** `defineLoader`, `defineAction`, `defineClientLoader`, `defineClientAction` **
+
+To get typesafety when defining loaders and actions, you can use the `defineLoader` and `defineAction` utilities:
+
+```ts
+import { defineLoader } from "@remix-run/node";
+
+export const loader = defineLoader(({ request }) => {
+  //                                  ^? Request
+});
+
+export const action = defineAction(({ context }) => {
+  //                                  ^? AppLoadContext
+});
+```
+
+Not only does this give you types for any arguments, but it also ensures you are returning single-fetch compatible types:
+
+```ts
+export const loader = defineLoader(() => {
+  return { hello: "world", badData: () => 1 };
+  //                       ^^^^^^^ Type error: `badData` is not serializable
+});
+
+export const action = defineAction(() => {
+  return { hello: "world", badData: new CustomType() };
```

Comment:
What if CustomType class has a toJSON?

Reply:
The contract for single-fetch (and `turbo-stream` which is what it is using under-the-hood) is that a type `T` in becomes a type `T` out. So `CustomType` -> `string` is not supported. If you want you can manually call `toJSON()`:

```ts
export let loader = () => {
  return { a: myCustomTypeInstance.toJSON() }
}
```

The point is to have less coercion and magic. Having the serialization be basically a glorified ID function is the thing that unlocks a bunch of the simplified types features. Plus it means we can support recursive types much better than the current approach.

Reply:
@sergiodxa FIY, I raised the same question on Discord: https://discord.com/channels/770287896669978684/1233404747885576293/1233417203693523104

Reply:
Cross posting my response from discord as well for visibility

> We're hesitant to add to much "custom" stuff like this to turbo-stream because it's a stopgap solution that will be replaced with the RSC wire streaming format once we land the RSC work.  So the limitations of what can and can't be streamed are going to eventually be defined by React, not us

I.e., if we add `toJson` support now, folks will start relying on it.  Then if we incorporate RSC and it doens't support `toJSON`, it would be a breaking change for your apps.  So we're keeping this simpler now until we know in greater detail how the RSC story shakes out.